### PR TITLE
[POC] DeepSeek V4 on RTX Pro 6000 (SM120)

### PR DESCRIPTION
**Pull request by jasl**
This is just a Proof-of-Concept PR made by GPT and me. If anyone can help get it to production, please take it.

ds

## Summary

This is a draft prototype for running DeepSeek V4 Flash on NVIDIA SM120/SM121
(GB10 / GeForce Blackwell-class) systems.

The goal is correctness-first enablement, not production performance. SM90/SM100
continue to use the existing optimized paths. SM120 uses reference fallbacks for
the DeepSeek V4 paths that currently fail with `Unsupported architecture`.

## Motivation

DeepSeek V4 depends on several kernels that are currently implemented for
SM90/SM100 only. On SM120/SM121, startup or first forward can fail in:

- DeepGEMM HyperConnection / mHC paths
- sparse MLA attention paths
- sparse attention indexer logits paths
- vendored DeepGEMM CMake arch filtering

This prototype is intended to unblock forward/generation on a DGX Spark / GB10
host while leaving room for later SM120-specific optimization.

## Approach

This patch adds an experimental SM120 reference path for DeepSeek V4:

- Allow vendored DeepGEMM to build for SM12x with CUDA 13.
- Pin the prototype DeepGEMM dependency that contains SM120 reference fallbacks.
- Add SM120 DeepSeek V4 reference attention controls.
- Reuse the existing DeepSeek V4 metadata / top-k / KV-cache flow where possible.
- Add a SM120 FP8 indexer logits fallback for the non-paged MQA logits path.
- Register the SM120 prototype environment variables to avoid startup warnings.

The reference path is intentionally conservative. It prioritizes matching the
DeepSeek V4 sparse attention semantics, including sparse indices and
sink-aware denominator behavior, over throughput.

## Known Limitations

- This is not an optimized SM120 backend.
- The reference paths are expected to be slow.
- The current validation target is DeepSeek V4 Flash with FP8 KV cache.
- Full 1M context is memory-heavy on this host; the tested command uses
  `--max-model-len 262144`.
- `/v1/completions` is raw text completion. For instruction-style testing,
  use `/v1/chat/completions` with `--tokenizer-mode deepseek_v4`.

## Build

Test host:

- NVIDIA DGX Spark / GB10
- SM120/SM121-class GPU
- CUDA 13
- vLLM dsv4 branch / DeepSeek V4 integration branch

Environment:

```bash
export PATH="/usr/local/cuda/bin:$PATH"
export CUDA_HOME="/usr/local/cuda"
export TRITON_PTXAS_PATH="/usr/local/cuda/bin/ptxas"
export CUDA_ARCH_LIST="120a"
export TORCH_CUDA_ARCH_LIST="12.0a"
```

Build vLLM:

```bash
cd ~/tmp/vllm
source ~/tmp/vllm/.venv/bin/activate

DEEPGEMM_SRC_DIR=~/tmp/DeepGEMM \
CCACHE_NOHASHDIR=true \
MAX_JOBS=64 \
pip install --verbose --no-build-isolation -e .
```

Expected build log should include:

```
DeepGEMM CUDA architectures: 12.0a
```

## Run

```
export PATH="/usr/local/cuda/bin:$PATH"
export CUDA_HOME="/usr/local/cuda"
export TRITON_PTXAS_PATH="/usr/local/cuda/bin/ptxas"
export CUDA_ARCH_LIST="120a"
export TORCH_CUDA_ARCH_LIST="12.0a"

export VLLM_SM120_REFERENCE_DEEPSEEK_V4_ATTENTION=1
export VLLM_SM120_REFERENCE_TOPK_CHUNK_SIZE=256
export VLLM_SM120_REFERENCE_QUERY_CHUNK_SIZE=128
```

Start server:

```
~/tmp/vllm/.venv/bin/vllm serve deepseek-ai/DeepSeek-V4-Flash \
  --trust-remote-code \
  --kv-cache-dtype fp8 \
  --block-size 256 \
  --tensor-parallel-size 2 \
  --max-model-len 262144 \
  --compilation-config '{"cudagraph_mode":"FULL_AND_PIECEWISE", "custom_ops":["all"]}' \
  --tokenizer-mode deepseek_v4 \
  --tool-call-parser deepseek_v4 \
  --enable-auto-tool-choice \
  --reasoning-parser deepseek_v4
```

Smoke Test

```
curl -sS --max-time 240 http://127.0.0.1:8000/v1/chat/completions \
  -H 'Content-Type: application/json' \
  -d '{
    "model": "deepseek-ai/DeepSeek-V4-Flash",
    "messages": [
      {"role": "user", "content": "Introduce yourself briefly."}
    ],
    "max_tokens": 128,
    "temperature": 0
  }'
```

Expected result: HTTP 200 with a normal assistant response.

### CODE REVIEW

**Comment by shahizat** [comment]
Hi @jasl, 

Thanks for supporting the SM12X community! I just tested using two Blackwell Maxq 6000 Pro GPUs, and it works fine. For BS1, it shows a reasonable inference speed(max peak was 37.8 tokens/s)

```
============ Serving Benchmark Result ============
Successful requests:                     100       
Failed requests:                         0         
Maximum request concurrency:             10        
Benchmark duration (s):                  137.52    
Total input tokens:                      12800     
Total generated tokens:                  12800     
Request throughput (req/s):              0.73      
Output token throughput (tok/s):         93.08     
Peak output token throughput (tok/s):    130.00    
Peak concurrent requests:                20.00     
Total token throughput (tok/s):          186.15    
---------------Time to First Token----------------
Mean TTFT (ms):                          3435.78   
Median TTFT (ms):                        3386.33   
P99 TTFT (ms):                           7097.29   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          81.22     
Median TPOT (ms):                        78.96     
P99 TPOT (ms):                           102.80    
---------------Inter-token Latency----------------
Mean ITL (ms):                           81.22     
Median ITL (ms):                         77.85     
P99 ITL (ms):                            81.95     
==================================================
```

@mgoin, @johnnynunez

**Comment by jasl** [comment]
Since the original DSv4 support PR is still drafting, I'm going to polish the essential kernels for now.
I will open a separate PR for that and work to make it mergeable.

**Comment by jasl** [comment]
Now

c=4, 1024/1024, 20 prompts：
graph default 0.900：98.29 tok/s
graph + --gpu-memory-utilization 0.925：98.10 tok/s

I chose concurrence 4 is becasue 1M context consumes about 7GB memory, for 2 x RTX Pro 6000, the remaining memory could serve 4 full context requests

**Comment by jasl** [comment]
My WIP branch https://github.com/jasl/vllm/tree/ds4-sm120

image

There is still much room to improve

**Comment by RaspberrySun** [comment]
> My WIP branch https://github.com/jasl/vllm/tree/ds4-sm120
> 
> image
> There is still much room to improve

I was wondering if this branch also supports DGX Spark with SM121 architecture, or if there are any known compatibility/performance issues when running it on SM121 hardware?

**Comment by jasl** [comment]
> > My WIP branch https://github.com/jasl/vllm/tree/ds4-sm120
> > image
> > There is still much room to improve
> 
> I was wondering if this branch also supports DGX Spark with SM121 architecture, or if there are any known compatibility/performance issues when running it on SM121 hardware?

Definitely support DGX Spark! You need 2 DGX Sparks to run the model, but I only have 1.
I've asked Johnny for help.

**Comment by jasl** [comment]
superceded by https://github.com/vllm-project/vllm/pull/40899
